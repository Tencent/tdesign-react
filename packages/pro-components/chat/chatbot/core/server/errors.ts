/* eslint-disable max-classes-per-file */
/**
 * Enhanced error classes for SSE client
 */

// 基础 SSE 错误类
export class SSEError extends Error {
  constructor(
    message: string,
    public readonly code: string,
    public readonly statusCode?: number,
    public readonly isRetryable: boolean = false,
    public readonly details?: any,
  ) {
    super(message);
    this.name = 'SSEError';
  }
}

// 连接错误
export class ConnectionError extends SSEError {
  constructor(message: string, statusCode?: number, details?: any) {
    super(message, 'CONNECTION_ERROR', statusCode, true, details);
    this.name = 'ConnectionError';
  }
}

// 超时错误
export class TimeoutError extends SSEError {
  constructor(details?: any, message = '请求超时') {
    super(message, 'TIMEOUT_ERROR', undefined, true, details);
    this.name = 'TimeoutError';
  }
}

// 解析错误
export class ParseError extends SSEError {
  constructor(message: string, details?: any) {
    super(message, 'PARSE_ERROR', undefined, false, details);
    this.name = 'ParseError';
  }
}

// 验证错误
export class ValidationError extends SSEError {
  constructor(message: string, details?: any) {
    super(message, 'VALIDATION_ERROR', undefined, false, details);
    this.name = 'ValidationError';
  }
}
