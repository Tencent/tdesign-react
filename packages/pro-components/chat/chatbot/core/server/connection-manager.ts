import { LoggerManager } from '../utils/logger';
import { TimeoutError } from './errors';
import { ConnectionInfo, ConnectionStats, SSEConnectionState } from './types';

/**
 * 连接管理器 - 直接使用全局Logger
 */
export class ConnectionManager {
  private connectionStartTime = 0;

  private stats: ConnectionStats;

  private logger = LoggerManager.getLogger(); // 直接使用全局Logger

  constructor(private connectionId: string) {
    this.stats = {};
  }

  /**
   * 处理连接错误
   */
  handleConnectionError(error: Error): boolean {
    this.logger.error(`Connection ${this.connectionId} error:`, error);
    this.stats.lastError = error;

    if (error instanceof TimeoutError) {
      this.logger.info('Timeout error occurred, no retry will be attempted');
    }

    this.cleanup();
    return false;
  }

  /**
   * 开始连接计时
   */
  startConnection(): void {
    this.connectionStartTime = Date.now();
    this.logger.debug(`Connection ${this.connectionId} started`);
  }

  /**
   * 连接成功
   */
  onConnectionSuccess(): void {
    const duration = Date.now() - this.connectionStartTime;
    this.logger.info(`Connection established in ${duration}ms`);
    this.stats.connectionTime = duration;
  }

  /**
   * 更新连接状态并记录统计信息
   */
  updateState(state: SSEConnectionState, error?: Error): void {
    if (error) {
      this.stats.lastError = error;
    }

    this.logger.debug(`State updated to ${state}`);
  }

  /**
   * 获取连接信息
   */
  getConnectionInfo(): ConnectionInfo {
    return {
      id: this.connectionId,
      retryCount: 0,
      lastActivity: Date.now(),
      stats: { ...this.stats },
    };
  }

  /**
   * 获取连接统计
   */
  getStats(): ConnectionStats {
    return { ...this.stats };
  }

  /**
   * 清理资源
   */
  cleanup(): void {
    this.stats = {};
    this.logger.debug(`Connection manager ${this.connectionId} cleaned up`);
  }
}
