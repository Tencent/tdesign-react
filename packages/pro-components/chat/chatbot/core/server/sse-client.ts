/* eslint-disable no-await-in-loop, max-classes-per-file */
import EventEmitter from '../utils/eventEmitter';
import { LoggerManager } from '../utils/logger';
import { ConnectionManager } from './connection-manager';
import { ConnectionError, TimeoutError } from './errors';
import { type SSEEvent, SSEParser } from './sse-parser';
import {
  type ConnectionInfo,
  DEFAULT_SSE_CONFIG,
  type SSEClientConfig,
  SSEConnectionState,
  type StateChangeEvent,
} from './types';

/**
 * SSE Client
 * 采用分层设计，分离了连接管理、状态管理、事件解析等职责
 */
export class SSEClient extends EventEmitter {
  public readonly connectionId: string;

  private state = SSEConnectionState.DISCONNECTED;

  private controller?: AbortController | null;

  private reader?: ReadableStreamDefaultReader<string>;

  private connectionManager: ConnectionManager;

  private parser: SSEParser;

  private timeoutTimer?: ReturnType<typeof setInterval>; // 统一超时定时器

  private config: SSEClientConfig;

  private logger = LoggerManager.getLogger();

  private url: string;

  private connectionInfo: ConnectionInfo;

  private firstTokenReceived = false;

  constructor(url: string) {
    super();
    this.url = url;
    this.connectionId = this.generateConnectionId();
    this.logger = LoggerManager.getLogger();
    this.connectionManager = new ConnectionManager(this.connectionId);

    // 初始化 SSE 解析器
    this.parser = new SSEParser();
    this.parser.onMessage = (event: SSEEvent) => {
      this.emit('message', event);
    };

    this.connectionInfo = {
      id: this.connectionId,
      url,
      state: this.state,
      createdAt: Date.now(),
      retryCount: 0,
      lastActivity: Date.now(),
      stats: {},
    };

    this.setupInternalEventHandlers();
  }

  /**
   * 连接 SSE 服务
   */
  async connect(config: SSEClientConfig): Promise<void> {
    if (this.state === SSEConnectionState.CONNECTED || this.state === SSEConnectionState.CONNECTING) {
      return;
    }

    this.config = {
      ...DEFAULT_SSE_CONFIG,
      ...config,
      headers: {
        ...DEFAULT_SSE_CONFIG.headers,
        ...config.headers,
      },
    };

    this.setState(SSEConnectionState.CONNECTING);
    this.connectionManager.startConnection();

    try {
      // 启动统一超时监控
      this.startTimeoutMonitor();

      await this.establishConnection();
      this.setState(SSEConnectionState.CONNECTED);
      this.connectionManager.onConnectionSuccess();
      await this.readStream();
    } catch (error) {
      this.handleConnectionError(error as Error);
    }
  }

  /**
   * 关闭连接
   */
  async abort(): Promise<void> {
    if (this.state === SSEConnectionState.DISCONNECTED || this.state === SSEConnectionState.CLOSING) {
      return;
    }

    this.setState(SSEConnectionState.CLOSING);

    try {
      if (this.reader) {
        await this.reader.cancel();
        this.reader = undefined;
      }

      if (this.controller && !this.controller.signal.aborted) {
        this.controller.abort();
      }

      this.connectionManager.cleanup();
      this.resetParser();
      this.emit('complete', true);
    } catch (error) {
      if (error.name !== 'AbortError') {
        this.logger.error('stream abort failed:', error);
        this.emit('error', error);
      }
    } finally {
      this.clearTimeouts();
      this.controller = null;
      this.setState(SSEConnectionState.CLOSED);
    }
  }

  /**
   * 获取连接状态
   */
  getStatus(): SSEConnectionState {
    return this.state;
  }

  /**
   * 获取连接信息
   */
  getInfo(): ConnectionInfo {
    return {
      ...this.connectionInfo,
      ...this.connectionManager.getConnectionInfo(),
    };
  }

  /**
   * 建立连接
   */
  private async establishConnection(): Promise<void> {
    this.controller = new AbortController();

    try {
      const response = await fetch(this.url, {
        ...this.config,
        signal: this.controller.signal,
        headers: {
          ...this.config.headers,
          Accept: 'text/event-stream',
          'Cache-Control': 'no-cache',
        },
      });

      if (!response.body || !response.ok) {
        this.emit(
          'error',
          new ConnectionError(`HTTP ${response.status}: ${response.statusText}`, response.status, response),
        );
        return;
      }
      this.reader = response.body.pipeThrough(new TextDecoderStream()).getReader();
    } catch (error: any) {
      if (error.name !== 'AbortError') {
        this.logger.error('sse request failed:', error);
        this.emit('error', error);
      }
    }
  }

  /**
   * 读取流数据
   */
  private async readStream(): Promise<void> {
    try {
      while (this.state === SSEConnectionState.CONNECTED && this.reader) {
        // eslint-disable no-await-in-loop
        const { done, value } = await this.reader.read();

        if (done) {
          this.logger.info(`Connection ${this.connectionId} stream ended normally`);
          this.emit('complete', false); // 发出流结束事件
          this.setState(SSEConnectionState.DISCONNECTED);
          this.clearTimeouts();
          return;
        }

        // 更新活动时间
        this.connectionInfo.lastActivity = Date.now();

        // 直接解析SSE数据
        this.parseSSEData(value);
      }
    } catch (error) {
      if (!this.controller?.signal.aborted) {
        this.logger.error(`Stream reading error for ${this.connectionId}:`, error);
        this.handleConnectionError(error as Error);
      } else {
        this.logger.debug(`Stream reading stopped for ${this.connectionId} (aborted)`);
      }
    }
  }

  /**
   * 启动统一超时监控
   */
  private startTimeoutMonitor(): void {
    if (!this.config.timeout || this.config.timeout <= 0) return;

    // 检查间隔为超时时间的一半，但不超过5秒
    const checkInterval = Math.min(this.config.timeout / 2, 5000);

    this.timeoutTimer = setInterval(() => {
      const now = Date.now();
      const timeSinceLastActivity = now - this.connectionInfo.lastActivity;

      if (timeSinceLastActivity >= this.config.timeout!) {
        // 根据当前状态决定超时类型
        if (this.state === SSEConnectionState.CONNECTING) {
          this.logger.warn(`Connection ${this.connectionId} establishment timeout after ${this.config.timeout}ms`);
          this.emit('error', new TimeoutError(`Connection establishment timed out after ${this.config.timeout}ms`));
        } else {
          this.logger.warn(`Connection ${this.connectionId} receive timeout after ${this.config.timeout}ms`);
          this.emit('error', new TimeoutError(`No data received for ${this.config.timeout}ms`));
        }

        this.abort(); // 触发连接中断
      }
    }, checkInterval);
  }

  /**
   * 解析SSE数据
   */
  private parseSSEData(chunk: string): void {
    // 使用独立的 SSE 解析器处理数据
    this.parser.parse(chunk);
    if (!this.firstTokenReceived) {
      this.firstTokenReceived = true;
      this.emit('start', chunk); // 派发start事件
    }
  }

  /**
   * 简化的错误处理
   */
  private handleConnectionError(error: Error) {
    this.connectionInfo.error = error;
    this.connectionManager.handleConnectionError(error);
    this.setState(SSEConnectionState.ERROR);
    this.emit('error', error);
  }

  /**
   * 清理超时定时器
   */
  private clearTimeouts(): void {
    if (this.timeoutTimer) {
      clearInterval(this.timeoutTimer);
      this.timeoutTimer = undefined;
    }
  }

  /**
   * 设置连接状态
   */
  private setState(newState: SSEConnectionState) {
    const oldState = this.state;
    this.state = newState;
    this.connectionInfo.state = newState;

    const stateChangeEvent: StateChangeEvent = {
      connectionId: this.connectionId,
      from: oldState,
      to: newState,
      timestamp: Date.now(),
    };

    this.emit('stateChange', stateChangeEvent);
    this.logger.debug(`Connection ${this.connectionId} state: ${oldState} -> ${newState}`);
  }

  /**
   * 重置解析器状态
   */
  private resetParser(): void {
    this.parser.reset();
  }

  /**
   * 设置内部事件处理器
   */
  private setupInternalEventHandlers(): void {
    this.on('error', (error) => {
      this.logger.error(`SSE Client ${this.connectionId} error:`, error);
    });

    this.on('complete', (isAborted) => {
      this.logger.info(`SSE Client ${this.connectionId} completed, aborted: ${isAborted}`);
    });
  }

  /**
   * 生成连接ID
   */
  private generateConnectionId(): string {
    // 使用this来访问实例属性，满足linter要求
    const timestamp = this.connectionInfo?.createdAt || Date.now();
    return `sse_${timestamp}_${Math.random().toString(36).substr(2, 9)}`;
  }
}
