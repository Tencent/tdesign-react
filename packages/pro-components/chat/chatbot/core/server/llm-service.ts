import type { AIMessageContent, ChatRequestParams, ChatServiceConfig, SSEChunkData } from '../type';
import { LoggerManager } from '../utils/logger';
import { BatchClient } from './batch-client';
import { SSEClient } from './sse-client';

// 与原有接口保持兼容
export interface ILLMService {
  /**
   * 处理批量请求（非流式）
   */
  handleBatchRequest(
    params: ChatRequestParams,
    config: ChatServiceConfig,
  ): Promise<AIMessageContent | AIMessageContent[]>;

  /**
   * 处理流式请求
   */
  handleStreamRequest(params: ChatRequestParams, config: ChatServiceConfig): Promise<void>;
}

/**
 * Enhanced LLM Service with error handling and connection management
 */
export class LLMService implements ILLMService {
  // 使用接口确保类型安全
  private sseClient: SSEClient;

  private batchClient: BatchClient | null = null;

  private isDestroyed = false;

  private logger = LoggerManager.getLogger();

  constructor(private config: ChatServiceConfig) {
    this.logger.info('Enhanced LLM Service initialized');
  }

  /**
   * 处理批量请求（非流式）
   */
  async handleBatchRequest(
    params: ChatRequestParams,
    config: ChatServiceConfig,
  ): Promise<AIMessageContent | AIMessageContent[]> {
    // 确保只有一个客户端实例
    this.batchClient = this.batchClient || new BatchClient();
    this.batchClient.on('error', (error) => {
      config.onError?.(error);
    });

    const req = (await config.onRequest?.(params)) || params;

    try {
      const data = await this.batchClient.request<AIMessageContent>(
        config.endpoint,
        {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            ...req.headers,
          },
          body: req.body,
        },
        config.timeout, // 现在timeout属性已存在
      );
      if (data) {
        return config.onComplete?.(false, req, data);
      }
    } catch (error) {
      config.onError?.(error);
      throw error;
    }
  }

  /**
   * 处理流式请求
   */
  async handleStreamRequest(params: ChatRequestParams, config: ChatServiceConfig): Promise<void> {
    if (!config.endpoint) return;
    this.sseClient = new SSEClient(config.endpoint);

    const req = (await config.onRequest?.(params)) || {};

    // 设置事件处理器
    this.sseClient.on('start', (chunk) => {
      config.onStart?.(chunk);
    });

    this.sseClient.on('message', (msg) => {
      config.onMessage?.(msg as SSEChunkData);
    });

    this.sseClient.on('error', (error) => {
      config.onError?.(error);
    });

    this.sseClient.on('complete', (isAborted) => {
      config.onComplete?.(isAborted, req);
    });

    await this.sseClient.connect(req);
  }

  /**
   * 关闭所有客户端连接
   */
  closeConnect(): void {
    if (this.sseClient) {
      this.sseClient.abort();
      this.sseClient = null;
    }
    if (this.batchClient) {
      this.batchClient?.abort();
      this.batchClient = null;
    }
  }

  /**
   * 获取连接统计
   */
  getSSEStats(): { id: string; status: string; info: any } | null {
    if (!this.sseClient) return null;

    return {
      id: this.sseClient.connectionId,
      status: this.sseClient.getStatus(),
      info: this.sseClient.getInfo(),
    };
  }

  /**
   * 销毁服务
   */
  async destroy(): Promise<void> {
    this.isDestroyed = true;
    this.closeConnect();
    this.logger.info('Enhanced LLM Service destroyed');
  }
}
