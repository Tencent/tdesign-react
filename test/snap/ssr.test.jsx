import glob from 'glob';
import MockDate from 'mockdate';
import { vi } from 'vitest';
import React from 'react';
import { renderToString } from 'react-dom/server';

// 固定时间，当使用 new Date() 时，返回固定时间，防止“当前时间”的副作用影响，导致 snapshot 变更，mockdate 插件见 https://github.com/boblauer/MockDate
MockDate.set('2020-12-28 00:00:00');

function runTest() {
  const files = glob.sync('src/**/_example/*.jsx', {
    ignore: ['src/watermark/_example/*.jsx'],
  });

  vi.mock('react-dom', async () => ({
    ...(await vi.importActual('react-dom')),
    createPortal: node => node
  }));
  
  describe('ssr snapshot test', () => {
    HTMLCanvasElement.prototype.getContext = vi.fn();

    files.forEach((file) => {
      it(`ssr test ${file}`, async () => {
        const demo = await import(`../../${file}`);
        const RealDemoComp = demo.default ? demo.default : demo;
        const html = renderToString(<RealDemoComp />);
        expect(html).toMatchSnapshot();
      });
    });
  });
}

runTest();
