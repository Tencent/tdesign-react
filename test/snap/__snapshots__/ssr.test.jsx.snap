// Vitest Snapshot v1

exports[`ssr snapshot test > ssr test src/icon/_example/Enhanced.jsx 1`] = `"<div><svg class=\\"t-icon cps-icon-home-sheep\\" style=\\"margin-right:8px\\"><use xlink:href=\\"#cps-icon-home-sheep\\"></use></svg><svg class=\\"t-icon cps-icon-home-sheep t-size-m\\" style=\\"margin-right:8px\\"><use xlink:href=\\"#cps-icon-home-sheep\\"></use></svg><svg class=\\"t-icon cps-icon-home-sheep t-size-l\\" style=\\"margin-right:8px\\"><use xlink:href=\\"#cps-icon-home-sheep\\"></use></svg><svg class=\\"t-icon cps-icon-home-sheep\\" style=\\"font-size:25px\\"><use xlink:href=\\"#cps-icon-home-sheep\\"></use></svg><br/><br/><svg class=\\"t-icon cps-icon-home-sheep\\" style=\\"color:red;margin-right:8px\\"><use xlink:href=\\"#cps-icon-home-sheep\\"></use></svg><svg class=\\"t-icon cps-icon-home-sheep\\" style=\\"color:green;margin-right:8px\\"><use xlink:href=\\"#cps-icon-home-sheep\\"></use></svg><svg class=\\"t-icon cps-icon-home-sheep\\" style=\\"color:orange;margin-right:8px\\"><use xlink:href=\\"#cps-icon-home-sheep\\"></use></svg><svg class=\\"t-icon t-icon-home\\"><use xlink:href=\\"#t-icon-home\\"></use></svg></div>"`;

exports[`ssr snapshot test > ssr test src/icon/_example/IconFontEnhanced.jsx 1`] = `"<div><i style=\\"margin-right:8px\\" class=\\"cps-icon cps-icon-home-sheep\\"></i><i style=\\"margin-right:8px\\" class=\\"cps-icon cps-icon-home-sheep t-size-m\\"></i><i style=\\"margin-right:8px\\" class=\\"cps-icon cps-icon-home-sheep t-size-l\\"></i><i style=\\"font-size:25px\\" class=\\"cps-icon cps-icon-home-sheep\\"></i><br/><br/><i style=\\"color:red;margin-right:8px\\" class=\\"cps-icon cps-icon-home-sheep\\"></i><i style=\\"color:green;margin-right:8px\\" class=\\"cps-icon cps-icon-home-sheep\\"></i><i style=\\"color:orange;margin-right:8px\\" class=\\"cps-icon cps-icon-home-sheep\\"></i><i class=\\"t-icon-home t-icon\\"></i></div>"`;
